okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java

```diff
@@ -159,26 +147,60 @@ private static ResponseBody buffer(MediaType mediaType, byte[] bodyBytes) {
 
     @Override
     public void enqueue(Callback callback) {
-        AsyncTracer tracer = new AsyncTracer("OkHttp: acquire-limiter");
-        ListenableFuture<Limiter.Listener> limiterListener = limiter.acquire();
-        request().tag(ConcurrencyLimiterListener.class).setLimiterListener(limiterListener);
-        Futures.addCallback(limiterListener, new FutureCallback<Limiter.Listener>() {
+        Futures.addCallback(executeAsync(), new FutureCallback<Response>() {
             @Override
-            public void onSuccess(Limiter.Listener listener) {
-                tracer.withTrace(() -> null);
-                enqueueInternal(callback);
+            public void onSuccess(Response response) {
+                try {
+                    callback.onResponse(RemotingOkHttpCall.this, response);
+                } catch (IOException e) {
+                    log.warn("Callback failure", e);
+                }
             }
 
             @Override
             public void onFailure(Throwable throwable) {
-                callback.onFailure(
-                        RemotingOkHttpCall.this,
-                        new IOException(new AssertionError("This should never happen, since it implies "
-                                + "we failed when using the concurrency limiter", throwable)));
+                if (throwable instanceof IOException) {
+                    callback.onFailure(RemotingOkHttpCall.this, (IOException) throwable);
+                } else {
+                    callback.onFailure(RemotingOkHttpCall.this,
+                            new IOException(
+                                    "This shouldn't happen - all throwables should be IOExceptions",
+                                    throwable));
+                }
             }
         }, MoreExecutors.directExecutor());
     }
 
+    private ListenableFuture<Response> executeAsync() {
+        AsyncTracer tracer = new AsyncTracer("OkHttp: acquire-limiter");
+        ListenableFuture<Limiter.Listener> limiterListener = limiter.acquire();
+        request().tag(ConcurrencyLimiterListener.class).setLimiterListener(limiterListener);
+        return Futures.transformAsync(limiterListener, listener -> {
+            tracer.withTrace(() -> null);
+            ListenableFutureCallback callback = new ListenableFutureCallback();
+            enqueueInternal(callback);
+            return callback;
+        }, MoreExecutors.directExecutor());
+    }
+
+    private final class ListenableFutureCallback extends AbstractFuture<Response> implements Callback {
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            RemotingOkHttpCall.this.cancel();
```

Comment:
nit: would prefer a static class which takes the RemotingOkHttpCall instance for readability